inc/features/blocks/personalization/edit.js

```diff
@@ -0,0 +1,351 @@
+ * usable markup.
+ */
+const ALLOWED_BLOCKS = [ 'altis/personalization-variant' ];
+
+/**
+ * Start with a default template of one variant.
+ */
+const TEMPLATE = [
+	[ 'altis/personalization-variant' ],
+];
+
+// Component for fetching and displaying the variant title string.
+const VariantTitle = ( { variant } ) => {
+	const audience = useSelect( select => {
+		return select( 'audience' ).getPost( variant.attributes.audience );
+	}, [ variant.attributes.audience ] );
+	const isLoading = useSelect( select => {
+		return select( 'audience' ).getIsLoading();
+	}, [] );
+
+	if ( variant.attributes.fallback ) {
+		return __( 'Fallback', 'altis-experiments' );
+	}
+
+	if ( ! variant.attributes.audience ) {
+		return __( 'Select audience', 'altis-experiments' );
+	}
+
+	if ( audience && audience.title && audience.title.rendered ) {
+		return audience.title.rendered;
+	}
+
+	if ( isLoading ) {
+		return __( 'Loading...', 'altis-experiments' );
+	}
+
+	return '';
+}
+
+// Audience picker input.
+const Edit = ( {
+	addVariant,
+	attributes,
+	className,
+	clientId,
+	copyVariant,
+	isSelected,
+	removeVariant,
+	setAttributes,
+	setVariantAttributes,
+	variants,
+} ) => {
+
+	// Track currently selected variant.
+	const defaultVariantClientId = ( variants.length > 0 && variants[ 0 ].clientId ) || null;
+	const [ activeVariant, setVariant ] = useState( defaultVariantClientId );
+
+	// Track the active variant index to show in the title.
+	const activeVariantIndex = variants.findIndex( variant => variant.clientId === activeVariant );
+
+	// Set clientId attribute if not set.
+	if ( ! attributes.clientId ) {
+		setAttributes( { clientId } );
+	}
+
+	// Ensure variant data is correct parentId is correct.
+	variants.forEach( variant => {
+		if ( ! variant.attributes.parentId || variant.attributes.parentId !== attributes.clientId ) {
+			setVariantAttributes( variant.clientId, {
+				parentId: attributes.clientId,
+			} );
+		}
+	} );
+
+	return (
+		<Fragment>
+			<BlockControls>
+				<Toolbar
+					className="altis-variants-toolbar"
+					controls={ [
+						{
+							icon: 'plus',
+							title: __( 'Add a variant', 'altis-experiments' ),
+							className: 'altis-add-variant-button',
+							onClick: () => {
+								const newClientId = addVariant();
+								setVariant( newClientId );
+							},
+						},
+					] }
+				>
+					{ variants.map( variant => (
+						<Button
+							key={ `variant-select-${ variant.clientId }` }
+							className={ `altis-variant-button components-icon-button has-text ${ activeVariant === variant.clientId && 'is-active' }` }
+							title={ __( 'Select variant', 'altis-experiments' ) }
+							onClick={ () => setVariant( variant.clientId ) }
+						>
+							<VariantTitle variant={ variant } />
+						</Button>
+					) ) }
+				</Toolbar>
+			</BlockControls>
+			<InspectorControls>
+				{ variants.map( variant => {
+					if ( variant.attributes.fallback ) {
```

Comment:
Since this is a complex function, should break out to a standalone component instead.